### DISCUSSION TITLE
Allow some auto imports to be relative and others absolute

### DISCUSSION THREAD

**Discussion by maxeonyx** [microsoft/pylance-release#6467] [Enhancement]
Hi - I have run into some frustrating behaviour with relative import suggestions. I hope this issue can be helpful.

When using a project that has multiple separate module hierarchies, it is impossible to configure the automatic imports in a way that is satisfactory.

The behaviour we want is that imports from the local module are imported relatively, and imports across modules are imported absolutely.

As far as I can tell, there is no way to get this behaviour, no matter what I set in `python.analysis.include`, `python.analysis.exclude` and `python.analysis.extraPaths`.

I should be able to specify which directories are expected on the python path at runtime, but this does not seem possible.

## Environment data

-   Pylance version: v2024.9.2
-   OS and version: Fedora Linux 40 Workstation
-   Python version: Python 3.12.6 (system provided)

## Code Snippet

![Screenshot from 2024-09-27 13-34-17](https://github.com/user-attachments/assets/b9d8afb0-e8c3-4cd6-a94b-15075dbdee6f)

A minimal example can be found here: https://github.com/maxeonyx/pyimport-issue-minimal

To run the project, use:

```python
PYTHONPATH=appsrc:depsrc python -m appmodule
```

## Repro Steps

1. Clone the following repo: https://github.com/maxeonyx/pyimport-issue-minimal

2. Comment out your VSCode user settings so that they don't interfere with the workspace settings in this project.

3. Open the repo in VSCode.

4. Open `appsrc/appmodule/__main__.py`. Take the suggested action. The created import is not correct

### Relative mode
The following imports should be inserted in `relative` import mode:

```py
from . import do_thing
from sharedmodule.patterns import retry
```

### Absolute mode
The following imports should be inserted in `absolute` import mode:

```py
from appmodule import do_thing
from sharedmodule.patterns import retry
```

## Actual behavior

The following imports are what are actually inserted:

### Relative mode:
```py
from . import do_thing
from ...depsrc.sharedmodule.patterns import retry
```
The `...depsrc.sharedmodule.patterns` import is NOT correct as per my `python.analysis.*` settings.

### Absolute mode
This works correctly:
```py
from appmodule import do_thing
from sharedmodule.patterns import retry
```

## Note

Note that `...depsrc` is NOT expected to be on the python path at runtime. I have set `python.analysis.autoSearchPaths` to `false` so the workspace root should *not* be one of the search paths.

If we have the `from ...depsrc.sharedmodule.patterns import retry` version, when we run with the following command, we will get `ImportError: attempted relative import beyond top-level package`.

```sh
PYTHONPATH=appsrc:depsrc python -m appmodule
```

**Comment by rchiodo** [comment]
Thanks for the issue. This would be an enhancement request, being able to specify which paths should be absolute and which are relative. Probably requiring a new setting.

**Reply by maxeonyx** [reply]
Thanks @rchiodo

**Comment by rchiodo** [comment]
Transferring to a discussion item or up votes.

**Comment by heejaechang** [comment]
@maxeonyx what you need to do is `exclude: ["./depsrc"]` and `extrapath: ["./depsrc"]`, that should make any file under `depsrc` as non-user file but `extrapath` make it accessible from your other folder as third party library.

but I found some issues around it so I opened a bug - https://github.com/microsoft/pylance-release/issues/6468

one thing. by doing this, `pylance` will stop indexing `depsrc` as user file so you need to use `packageIndexDepths` setting to control how much we want from it.

**Reply by maxeonyx** [reply]
@heejaechang Thanks - I have tried that actually, but as you say pylance then stops indexing it. In my example all the files in `sharedmodule` are user files, so I see going this direction is a workaround. Thanks for looking into this, though.

**Reply by maxeonyx** [reply]
@heejaechang After thinking some more, I think you are right and @rchiodo I think this doesn't necessarily need to be a feature suggestion.

Already, when I set `"python.analysis.importFormat": "relative"`, some imports are still absolute. Standard library and third-party imports are still absolute. The problem is just that I don't seem able to configure some of my own code to behave in that way too (without losing type checking / indexing on it).

**Comment by maxeonyx** [comment]
I think there is two separate concepts here, but the current way that Pylance's settings are structured makes it hard to separate them:
- A) "Relative import possible" - which files are (expected to be) part of the same top-level python module at runtime? Which files *can* use relative imports to each other, without getting a relative import error?
- B) "User files vs non-user files" - should the file be indexed & type-checked, or only indexed "when necessary" to provide completions and type hints?

**Comment by maxeonyx** [comment]
@Tinche has a very concise screenshot demonstrating the issue here https://github.com/microsoft/pylance-release/discussions/4802#discussioncomment-11192866

I also don't think this should be considered an enhancement.

**Comment by karpetrosyan** [comment]
I think the import format setting is currently broken, so it's more of a bugfix than a feature. You can’t really use that setting because it generates invalid imports when importing from `tests/` folder. Tried a bunch of workarounds, none worked. Would be great to have a fix

**Reply by rchiodo** [reply]
Not sure what the fix would be. How would we know which folders should be relative and which ones shouldn't? I think the only way would be for you to exclude files as being relative?

**Reply by maxeonyx** [reply]
To start with, by looking at the apparent sys.path

```
repo
├── src1
│   ├── modfoo
│   │   ├── __init__.py
│   │   ├── A.py
│   │   └── B.py
│   └── modbar
│       ├── __init__.py
│       └── C.py
└── src2
    └── modbaz
        ├── __init__.py
        └── D.py
```

~~~
PYTHONPATH=src1:src2
~~~

We have some easy cases!

If I am in file `src1/modfoo/A.py` and I import a symbol from `src2/modbaz/D.py` then this must be an absolute import as they come from different entries on `sys.path` **and** they have different module names (filename / directory name / `.so` file name) under their respective `sys.path` entry.

**This currently doesn't work** - but this is the one that most people are reporting (including me).

It is true that python path entries can be overlapping but that is generally considered degenerate and can be ignored.

Then we get to the (somewhat) harder case:

I am in `src1/modfoo/A.py` and import a symbol from `src1/modbar/C.py`

This must also be an absolute import, as while the files are under the same entry on `sys.path`, they are in a different top-level module.

Finally, if I am in `src1/modfoo/A.py` and import from `src1/modfoo/B.py`, then only now can it finally be a relative import: both the current module and the target symbol are under the same `sys.path` entry and in a top level module of the same name.

**Reply by karpetrosyan** [reply]
Ah, I think you just described the fix I had in mind. Simply put, check if the generated import path goes outside the sys paths, and if it does, treat it as invalid and use an absolute import.

**Reply by rchiodo** [reply]
> Ah, I think you just described the fix I had in mind. Simply put, check if the generated import path goes outside the sys paths, and if it does, treat it as invalid and use an absolute import.

This sounds doable. Things crossing sys.paths should not be relative.

The other idea sounds like a matter of opinion. If there in the same sys path, we can't tell if you intended them to not be relative if you told us to use relative paths.

**Reply by maxeonyx** [reply]
To be clear, in the second case (`modfoo` vs. `modbar` in the same `src1` sys.path entry) a relative import still won't work, as this would cross a top level module boundary. src1 is just a directory (containing modules) and that's what gets added to sys.path 

But I agree this is harder and I think it's maybe less common.
